samples/msal-react-sample/src/msal-react/index.tsx

```diff
@@ -1,82 +1,87 @@
+    return useContext(MsalContext);
+};
+
+export function useHandleRedirect(): TokenResponse | null {
+    const msal = useMsal();
     const [ redirectResponse, setRedirectResponse ] = useState<TokenResponse | null>(null);
 
     useEffect(() => {
-        context?.handleRedirectPromise()
+        msal.handleRedirectPromise()
             .then(response => setRedirectResponse(response));
             // TODO: error handling
-    }, []);
+    }, [msal]);
 
-    return [ redirectResponse ];
+    return redirectResponse;
 }
 
-export const withMsal = () => (C:React.ComponentType) => (props:any) => {
-    return (
-        <MsalConsumer>
-            {msal => (
-                <C
-                    {...props}
-                    msal={msal}
-                />
-            )}
-        </MsalConsumer>
-    )
+// References for HOC pattern:
+//  https://react-typescript-cheatsheet.netlify.app/docs/hoc/intro
+//  https://www.pluralsight.com/guides/higher-order-composition-typescript-react
+export const withMsal = <P extends IMsalProps = IMsalProps>(Component: React.ComponentType<P>) => {
+    const ComponentWithMsal: React.FunctionComponent<P> = (props) => {
+        const msal = useMsal();
+        return <Component {...props} {...msal} />;
+    };
+
+    const componentName = Component.displayName || Component.name || "Component";
+    ComponentWithMsal.displayName = `withMsal(${componentName})`;
+
+    return ComponentWithMsal;
+};
+
+type MsalAuthenticationHook = {
+    isAuthenticated: boolean,
+    error: Error | null,
+    msal: IMsalProps
 }
 
-export function useIsAuthenticated(request: AuthenticationParameters = {}, forceLogin: boolean = false, usePopup: boolean = true): [ boolean, Error | null ] {
-    const context = useContext(MsalContext);
-    const authenticated = !!context?.getAccount();
+// TODO: Should `forceLogin` and `authenticationMethod` be contained in an object so we can add to the parameters without breaking changes to the API?
+export function useMsalAuthentication(request: AuthenticationParameters = {}, forceLogin: boolean = false, authenticationMethod: AuthenticationMethod = AuthenticationMethod.POPUP): MsalAuthenticationHook {
+    const msal = useMsal();
+    const isAuthenticated = !!msal.account;
 
+    // TODO: This state will only exist in this hook, may be safer to move the error state into the MsalContext
     const [ error, setError ] = useState<Error | null>(null);
 
-    async function loginInteractively(): Promise<void> {
-        if (usePopup) {
-            context?.loginPopup(request)
-            .catch((error) => {
-                setError(error);
-            });
-        } else {
-            context?.loginRedirect(request);
-        }
-        
-    }
+    // TODO: This is error prone because it asynchronously sets state, but the component may be unmounted before the process completes
+    //  May be better to move this into the the MsalContext
+    const loginInteractively = useCallback(
```

Comment:
I personally am not a fan of the design where logic around forcing login and interaction is part of configuring MSAL itself, which is why we the first version put that logic into `AuthenticatedComponent`. As an ethos, we try to to favor composition over configuration, and delegating requiring logic, preferred interaction, etc to each component tree that requires authentication more closely aligns with that, I believe.